metals/src/main/scala/scala/meta/internal/metals/testProvider/TestSuitesProvider.scala

```diff
@@ -1,165 +1,337 @@
+            buildTargetUpdates: _*
+          )
+        )
+      }
     }
     else Future.successful(())
 
+  /**
+   * BSP server return fully qualified names of all test suites per build target
+   * Remove cached entries which were deleted (not returned by BSP)
+   */
+  private def removeStaleTestSuites(
+      symbolsPerTargets: List[SymbolsPerTarget]
+  ): Map[BuildTarget, Seq[TestExplorerEvent]] = {
+    // when test suite is deleted it has to be removed from cache
+    symbolsPerTargets.map { case SymbolsPerTarget(buildTarget, testSymbols) =>
+      val fromBSP = testSymbols.values.toSet.map(FullyQualifiedName(_))
+      val cached = index.getSuiteNames(buildTarget)
+      val diff = (cached -- fromBSP)
+      val removed = diff.foldLeft(List.empty[TestExplorerEvent]) {
+        case (deleted, unusedClassName) =>
+          index.remove(buildTarget, unusedClassName) match {
+            case Some(entry) => entry.testClass.asRemove :: deleted
+            case None => deleted
+          }
+      }
+      (buildTarget, removed)
+    }.toMap
+  }
+
+  /**
+   * Returns discovered test entries per build target.
+   */
+  private def getTestEntries(
+      symbolsPerTarget: Seq[SymbolsPerTarget]
+  ): Map[BuildTarget, Seq[(TestEntry, TextDocument)]] = {
+    val entries = for {
+      SymbolsPerTarget(buildTarget, testSymbols) <- symbolsPerTarget
+      cachedSuites = index.getSuites(buildTarget)
+      (symbol, fullyQualifiedClassName) <- testSymbols
+        .readOnlySnapshot()
+        .toList
+      // IMPORTANT this check is meant to check for class name, not a symbol
+      fullyQualifiedName = FullyQualifiedName(fullyQualifiedClassName)
+      if !cachedSuites.contains(fullyQualifiedName)
+      entryWithDoc <- {
+        val mSymbol = mtags.Symbol(symbol)
+        computeTestEntry(
+          buildTarget,
+          mSymbol,
+          fullyQualifiedName
+        )
+      }
+    } yield entryWithDoc
+    entries.groupBy(_._1.buildTarget)
+  }
+
+  /**
+   * Compute BuildTargetUpdates from added and deleted entries.
+   * For added entry check if it's located in currently opened file, if yes try to find test cases for it.
+   */
+  private def getBuildTargetUpdates(
+      deletedSuites: Map[BuildTarget, Seq[TestExplorerEvent]],
+      addedSuites: Map[BuildTarget, Seq[TestExplorerEvent]],
+      addedTestCases: Map[BuildTarget, Seq[TestExplorerEvent]]
+  ): Seq[BuildTargetUpdate] = {
+
+    val buildTargetUpdates =
+      (deletedSuites.keySet ++ addedSuites.keySet ++ addedTestCases.keySet)
+        .map { buildTarget =>
+          val deleted = deletedSuites.get(buildTarget).getOrElse(Seq.empty)
+          val added = addedSuites.get(buildTarget).getOrElse(Seq.empty)
+          val added0 = addedTestCases.get(buildTarget).getOrElse(Seq.empty)
+          BuildTargetUpdate(buildTarget, deleted ++ added ++ added0)
+        }
+        .filterNot(_.events.isEmpty())
+        .toSeq
```

Comment:
Hmm, it works but looks unusual for me. Maybe smth like:
```scala
val aggregated = 
  (deletedSuites.toSeq ++ addedSuites.toSeq ++ addedTestCases.toSeq)
    .foldLeft(Map.empty[BuildTarget, List[TestExplorerEvent]]){
      case (acc, (target, events)) =>
         val prev = acc.getOrElse(target, Nil)
         acc.updated(target, events ::: prev)
    }

aggregated.map {case (target, events) => BuildTargetUpdate(target, events)}
```

Reply:
Nice suggestion, thanks!